firestore/src/main/java/com/firebase/ui/firestore/paging/FirestorePagingAdapter.java

```diff
@@ -0,0 +1,118 @@
+
+    private final Observer<LoadingState> mStateObserver =
+            new Observer<LoadingState>() {
+                @Override
+                public void onChanged(@Nullable LoadingState state) {
+                    if (state == null) {
+                        return;
+                    }
+
+                    onLoadingStateChanged(state);
+                }
+            };
+
+    private final Observer<PagedList<DocumentSnapshot>> mDataObserver =
+            new Observer<PagedList<DocumentSnapshot>>() {
+                @Override
+                public void onChanged(@Nullable PagedList<DocumentSnapshot> snapshots) {
+                    if (snapshots == null) {
+                        return;
+                    }
+
+                    onListChanged(snapshots);
+                }
+            };
+
+    public static class DiffCallback<T> extends DiffUtil.ItemCallback<DocumentSnapshot> {
+
+        private final SnapshotParser<T> mParser;
+
+        public DiffCallback(SnapshotParser<T> parser) {
+            mParser = parser;
+        }
+
+        @Override
+        public boolean areItemsTheSame(DocumentSnapshot oldItem, DocumentSnapshot newItem) {
+            return oldItem.getId().equals(newItem.getId());
+        }
+
+        @Override
+        public boolean areContentsTheSame(DocumentSnapshot oldItem, DocumentSnapshot newItem) {
+            T oldModel = mParser.parseSnapshot(oldItem);
+            T newModel = mParser.parseSnapshot(newItem);
+
+            return oldModel.equals(newModel);
+        }
+    }
+
+    public FirestorePagingAdapter(@NonNull FirestorePagingOptions<T> options) {
+        super(new DiffCallback<>(options.getParser()));
+
+        mParser = options.getParser();
+        mData = options.getData();
+
+        if (options.getOwner() != null) {
+            options.getOwner().getLifecycle().addObserver(this);
+        }
+    }
+
+    @OnLifecycleEvent(Lifecycle.Event.ON_START)
+    public void startListening() {
+        mData.getSnapshots().observeForever(mDataObserver);
+        mData.getLoadingState().observeForever(mStateObserver);
+    }
+
+    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+    public void stopListening() {
+        mData.getSnapshots().removeObserver(mDataObserver);
+        mData.getLoadingState().removeObserver(mStateObserver);
+    }
+
+    @NonNull
+    @Override
+    public abstract VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType);
+
+    @Override
+    public void onBindViewHolder(@NonNull VH holder, int position) {
+        DocumentSnapshot snapshot = getItem(position);
+        onBindViewHolder(holder, position, mParser.parseSnapshot(snapshot));
+    }
+
+    protected abstract void onBindViewHolder(@NonNull VH holder, int position, T model);
+
+    protected void onLoadingStateChanged(@NonNull LoadingState state) {
+        // For overriding
+    }
+
+    private void onListChanged(@NonNull PagedList<DocumentSnapshot> snapshots) {
```

Comment:
Couldn't we just inline this method? It just forwards a param to another method and confused me...